lib/Dialect/Torch/IR/TorchOps.cpp

```diff
@@ -763,6 +787,58 @@ void AtenLenTOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
   });
 }
 
+//===----------------------------------------------------------------------===//
+// AtenAddTensorOp
+//===----------------------------------------------------------------------===//
+
+void AtenAddTensorOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+                                                  MLIRContext *context) {
+  patterns.add(+[](AtenAddTensorOp op, PatternRewriter &rewriter) {
+    // The lhs and rhs of the add.tensor op should be 0d tensors for the
+    // canonicalization to be carried out.
+    // `aten.add.tensor(self, other, alpha)` is canonicalized to
+    // `aten.add.int(self, aten.mul.int(other, alpha))`.
+    Operation *lhsValueTensorLiteralOp =
+        op.self().getDefiningOp<ValueTensorLiteralOp>();
+    Operation *lhsPrimNumToTensorScalarOp =
+        op.self().getDefiningOp<PrimNumToTensorScalarOp>();
+    llvm::Optional<Value> lhs;
+    if (lhsValueTensorLiteralOp) {
+      lhs = getScalarValue(lhsValueTensorLiteralOp, op.getLoc(), rewriter);
+    } else if (lhsPrimNumToTensorScalarOp)
+      lhs = getScalarValue(lhsPrimNumToTensorScalarOp, op.getLoc(), rewriter);
+    else
+      return failure();
+
+    llvm::Optional<Value> rhs;
+    Operation *rhsValueTensorLiteralOp =
+        op.other().getDefiningOp<ValueTensorLiteralOp>();
+    Operation *rhsPrimNumToTensorScalarOp =
+        op.other().getDefiningOp<PrimNumToTensorScalarOp>();
+    if (rhsValueTensorLiteralOp)
+      rhs = getScalarValue(rhsValueTensorLiteralOp, op.getLoc(), rewriter);
+    else if (rhsPrimNumToTensorScalarOp) {
+      rhs = getScalarValue(rhsPrimNumToTensorScalarOp, op.getLoc(), rewriter);
+    } else
+      return failure();
```

Comment:
nit: This is a bit repetitive. I think it would be better to have `getScalarValue` take a `Value input` as the argument rather than an `Operation *`. Then you can replace this with:

```cpp
Value lhs = getScalarValue(op.self(), op.getLoc(), rewriter);
Value rhs = getScalarValue(op.other(), op.getLoc(), rewriter);

if (!lhs || !rhs)
  return failure();
```

Note that there is no need to return an `llvm::Optional`, since `Value` can be `null`.

Then in `getScalarValue` you would just do:

```cpp
if (auto valueTensorLiteralOp = input.getDefiningOp<ValueTensorLiteralOp>()) {
} else if (auto primNumToTensorScalarOp = input.getDefiningOp<PrimNumToTensorScalarOp>()) {
}
```

Reply:
I refactored based on the directions in here